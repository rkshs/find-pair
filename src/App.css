.App {
  text-align: center;
}

.grid {
  margin-top: 20px;
  display: grid;
  grid-template-columns: 120px 120px 120px 120px 120px 120px;
  grid-gap: 20px;
}


