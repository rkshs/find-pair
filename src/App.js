import {useEffect, useState, useRef} from 'react';
import Card from './components/Card'
import Timer from './components/Timer'
import './App.css';

 const cardImages = [
     {'className': 'card-image-1', matched: false},
     {'className': 'card-image-2', matched: false},
     {'className': 'card-image-3', matched: false},
     {'className': 'card-image-4', matched: false},
     {'className': 'card-image-5', matched: false},
     {'className': 'card-image-6', matched: false},
     {'className': 'card-image-7', matched: false},
     {'className': 'card-image-8', matched: false},
     {'className': 'card-image-9', matched: false},
     {'className': 'card-image-10', matched: false},
     {'className': 'card-image-11', matched: false},
     {'className': 'card-image-12', matched: false},
     {'className': 'card-image-13', matched: false},
     {'className': 'card-image-14', matched: false},
     {'className': 'card-image-15', matched: false},
     {'className': 'card-image-16', matched: false},
     {'className': 'card-image-17', matched: false},
     {'className': 'card-image-18', matched: false},
 ]

function App() {

    const [cards, setBunchCards] = useState([]);
    const [turn, setTurn] = useState(0);
    const [chosenOne, setChosenOne] = useState(null);
    const [chosenTwo, setChosenTwo] = useState(null);
    const [disabled, setDisabled] = useState(false);
    const [timer, setTimer] = useState({
        started: false,
        reset: false
    });

    const timerRef = useRef(null);


    const shuffle = () => {
    const newBunchCards = [...cardImages, ...cardImages]
        .sort(() => Math.random() - 0.5)
        .map((el) => {
        return {...el, id: Math.random()}
    })

        setChosenOne(null);
        setChosenTwo(null);
        setBunchCards(newBunchCards);
        setTurn(0)
        timerRef.current.resetTimer();
    }

    const setChoose = (card) => {
        if (!timer.started) {
            setTimer(() => ({started: true}));
        }
        chosenOne ? setChosenTwo(card) : setChosenOne(card);
    }

    useEffect(() => {
        if (chosenOne && chosenTwo) {
            setDisabled(true);

            if (chosenOne.className === chosenTwo.className) {
                setBunchCards(newBunch => {
                    return newBunch.map(card => {
                        if (card.className === chosenOne.className) {
                            return {...card, matched: true}
                        } else {
                            return card
                        }
                    })
                })
                reseter();
            } else {
                setTimeout(() => reseter(), 1500);
            }
        }
    }, [chosenOne, chosenTwo])

    const reseter = () => {
        setChosenOne(null);
        setChosenTwo(null);
        setTurn(t => t + 1);
        setDisabled(false);
    }

    useEffect(() => {
        shuffle();
    }, [])

    return (
    <div className="App">
        <div className="container">
            <h1>Find a pair game</h1>
            <div className="counters">
                <span>Turns: {turn}</span>
                <button onClick={shuffle}>New game</button>
                <span>
                    Time:
                    <Timer
                        timerStarted={timer.started}
                        ref={timerRef}
                    />
                </span>
            </div>
            <div className='grid'>
                {
                    cards.map(card => (
                        <Card
                            key={card.id}
                            card={card}
                            setChoose={setChoose}
                            flip={card === chosenOne || card === chosenTwo || card.matched}
                            disabled={disabled}
                        />
                    ))
                }
            </div>
        </div>
    </div>
    );
}

export default App;
