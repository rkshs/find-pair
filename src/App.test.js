import { render } from '@testing-library/react';
import '@testing-library/jest-dom'
import App from './App';

describe('main App component', () => {
    test('match snapshot', () => {
        const { container } = render(<App />);

        expect(container).toMatchSnapshot();
    });
});
