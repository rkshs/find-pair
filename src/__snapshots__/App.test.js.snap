// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main App component match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="container"
    >
      <h1>
        Find a pair game
      </h1>
      <div
        class="counters"
      >
        <span>
          Turns: 
          0
        </span>
        <button>
          New game
        </button>
        <span>
          Time:
          <span
            class="time"
          >
            <span
              class="minute"
            >
              00
            </span>
            <span>
              :
            </span>
            <span
              class="second"
            >
              00
            </span>
          </span>
        </span>
      </div>
      <div
        class="grid"
      >
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-11"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-13"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-14"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-16"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-8"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-6"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-8"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-18"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-1"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-9"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-6"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-2"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-9"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-7"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-13"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-3"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-11"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-5"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-17"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-15"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-1"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-12"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-15"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-2"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-14"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-7"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-3"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-17"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-12"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-10"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-4"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-18"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-4"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-16"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-5"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-10"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="container"
    >
      <h1>
        Find a pair game
      </h1>
      <div
        class="counters"
      >
        <span>
          Turns: 
          0
        </span>
        <button>
          New game
        </button>
        <span>
          Time:
          <span
            class="time"
          >
            <span
              class="minute"
            >
              00
            </span>
            <span>
              :
            </span>
            <span
              class="second"
            >
              00
            </span>
          </span>
        </span>
      </div>
      <div
        class="grid"
      >
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-11"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-11"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-14"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-9"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-3"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-9"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-1"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-8"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-4"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-3"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-7"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-17"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-18"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-15"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-16"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-16"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-4"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-6"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-6"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-12"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-14"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-18"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-15"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-2"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-17"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-7"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-2"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-13"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-13"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-5"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-5"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-1"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-10"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-8"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-12"
          />
        </div>
        <div
          class="card  "
        >
          <div
            class="card-cover"
          />
          <div
            class="card-image card-image-10"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
